touchforms/backend/persistence.py

```diff
@@ -24,32 +26,114 @@ def restore(sess_id, factory, override_state=None):
         state.update(override_state)
     return factory(**state)
 
-# TODO integrate with real caching framework (django ideally)
-def cache_set(key, value, timeout):
-    with open(cache_path(key), 'w') as f:
-        f.write(json.dumps(value).encode('utf8'))
 
-# TODO integrate with real caching framework (django ideally)
+def cache_set(key, value):
+    if key is None:
+        raise KeyError
+    postgres_insert(key, value)
+
+
 def cache_get(key):
+    if key is None:
+        raise KeyError
+    try:
+        return postgres_lookup(key)
+    except KeyError:
+        return cache_get_old(key)
+
+
+def cache_del(key):
+    conn = get_conn()
+    cursor = conn.cursor()
+    postgres_delete(cursor, key)
+    conn.commit()
+    conn.close()
+
+
+def postgres_select(cursor, key):
+    sel_sql = replace_table("SELECT * FROM %(kwarg)s WHERE sess_id=?")
+    sel_params = [str(key)]
+    cursor.execute(sel_sql, sel_params)
+
+
+def postgres_delete(cursor, key):
+    postgres_select(cursor, key)
+    if cursor.rowcount is not 0:
+        del_sql = replace_table("DELETE FROM %(kwarg)s WHERE sess_id=?")
+        del_params = [str(key)]
+        cursor.execute(del_sql, del_params)
+
+
+def postgres_lookup(key):
+    conn = get_conn()
+    cursor = conn.cursor()
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount is 0:
+        raise KeyError
+    value = cursor.fetchone()[1]
+
+    conn.close()
+    jsonobj = json.loads(value.decode('utf8'))
+    return jsonobj
+
+
+def postgres_insert(key, value):
+    conn = get_conn()
+    cursor = conn.cursor()
+
+    postgres_delete(cursor, key)
+
+    ins_sql = replace_table("INSERT INTO %(kwarg)s (sess_id, sess_json) VALUES (?, ?)")
+    ins_params = [str(key), json.dumps(value).encode('utf8')]
+
+    cursor.execute(ins_sql, ins_params)
+
+    conn.commit()
+    conn.close()
+
+
+def get_conn():
```

Comment:
so if we figure out connection pooling this would like return a statically defined / global object